QuantLib/ql/math/sampledcurve.cpp

```diff
@@ -30,6 +30,19 @@ namespace QuantLib {
             return (values_[jmid]+values_[jmid-1])/2.0;
     }
 
+    SampledCurve SampledCurve::derivative() const {
+        QL_REQUIRE(size()>=2,
```

Comment:
Instead of using a different grid, you could use the same grid as the values and use the central derivative `(values_[i+1]-values_[i-1])/(grid_[i+1]-grid_[i-1])`.